### PR TITLE
Java frontend: create static initialiser support functions earlier

### PR DESCRIPTION
**Pull request by smowton**
This creates the global variables that indicate whether a static initialiser
has already run during java_bytecode_languaget::typecheck, rather than introducing
them during method conversion as before.

This means that symex-driven lazy loading will be able to rely on the globals being
mentioned in __CPROVER_initialize, even if the first method to reference the static
initialiser is converted after _initialize has been created.

### CODE REVIEW

**Comment by smowton** [comment]
@thk123 @mgudemann @tautschnig @peterschrammel this is now rebased and ready to review

File: src/java_bytecode/java_static_initializers.cpp
Comment (thk123) [inline src/java_bytecode/java_static_initializers.cpp]: This probably shouldn't be checked in

File: src/java_bytecode/java_static_initializers.cpp
Comment (thk123) [inline src/java_bytecode/java_static_initializers.cpp]: Otherwise to what? If this follows on from the function comment it should probably be part of it

File: src/java_bytecode/java_static_initializers.cpp
Comment (thk123) [inline src/java_bytecode/java_static_initializers.cpp]: `lookup_reference` should be used here

File: src/java_bytecode/java_static_initializers.cpp
Comment (thk123) [inline src/java_bytecode/java_static_initializers.cpp]: @svorenova - you should have a look at this and ensure your change for generic base classes won't interfere with this (And vice versa)

**Review by thk123 — APPROVED** [review]
A please to review this: well documented &  well structured.

**Comment by smowton** [comment]
TG bump: https://github.com/diffblue/test-gen/pull/1438

**Review by thk123 — DISMISSED** [review]
Which makes me all the more sad to block until the TG pointer bump is passing CI

**Comment by smowton** [comment]
@thk123 changes applied

File: src/java_bytecode/java_static_initializers.cpp
Comment (thk123) [inline src/java_bytecode/java_static_initializers.cpp]: `Ssume`?

Comment (smowton) [reply src/java_bytecode/java_static_initializers.cpp]: gah. fixed.

File: src/java_bytecode/java_static_initializers.cpp
Comment (majakusber) [reply src/java_bytecode/java_static_initializers.cpp]: While the PR for my changes is still awaiting approval, I'm fairly confident at this point that there should be no interference. I will keep this in mind though until the PR is merged.

**Comment by smowton** [comment]
@thk123 this is passing now

**Comment by smowton** [comment]
@mgudemann @peterschrammel this is ready for review

**Review by thk123 — APPROVED** [review]
Remembered I had actually reviewed this - so restoring my original approval now TG is working

**Comment by smowton** [comment]
@chrisr-diffblue could also review?

**Review by chrisr-diffblue — APPROVED** [review]
Looks good to me, so I'll approve - but I do wonder why the 9573ad9 is included in this PR as it seems an entirely independent issue? As it's still a separate commit I'm not going to block this though.

**Comment by smowton** [comment]
@chrisr-diffblue that's a dependent PR that will be merged before this one

**Comment by smowton** [comment]
Rebased now the dependent PR is in.

**Pull request merged**
